### ISSUE TITLE
Construct A Modular Architecture

### ISSUE THREAD

**Issue by MrM0nkey** [nils-wisiol/pypuf#13]
Design a modular architecture for this project. In terms of modular it must be easy to add new puf simulations and learning algorithms in order to run experiments.
There some components which should be considered:
* PUF simulation
* Learning algorithms
* Experiments
* Results
* Process pool
* Command line interface

It should not be hard to reuse a lot of existing code for that.

**Comment by MrM0nkey** [comment]
Here is a first prototype of an [architekture](https://cacoo.com/diagrams/ej0IjbAIwRJOV9Hd#C38C9).
I think thats it. Simulation, Learner and Result are abstract classes, where every Learner needs his own Result class. It is possible to code a universal Result class which is generated by a list of learner attributes, but a specialized Result for each Learner seems cleaner. Simulation is a interface for different puf simulation. Every simulation must provide a TrainingSet with challenges, responses and the size of challenges and responses in oder to be learned by a Learner. The possibility to run multiple Experiments in parallel is supported by the Experimenter Class. The Experiment Class should be able to log the result into a File. The CMDLine Class the Interface between the command line script and the Experimenter.
CMDLine should read the experiment parameters from a file and starts the Experimenter.

**Comment by MrM0nkey** [comment]
In order to use Python 3 multiprocessing it is necessary to Instantiate the Experiments in the work Method for the process pool.

**Comment by MrM0nkey** [comment]
# Problem 1
I tried to implement the Experiment Class and a difficulty occurred.
How do we want to integrate new Simulations? I mean how should we determine which learner class and simulation class to instantiate.
It should be possible to define a couple of experiments in a file which the CMDLine can read.
For example:
```
learner_module_name  simulation_module_name  n  k  crp_count seed    restarts
learner                         simulation                        16 4 120000     0xbeef 6
learner                         simulation                        16 4 120000     0xbeef 6
                                          .
                                          .
                                          . 
```
I have two ideas to solve this problem:

- hardcode a dictionary in order to map names to classes
- load classes dynamic
Code for the second solution:
``` python
  name = "learner"
  learner_module = importlib.import_module(name)
  learner_class = getattr(learner_module, name.title())
```
# Problem 2
Another problem is that the LTF simulation allows "arbitrary" input, output transformation and a combiner function. Each combination of these parameter is a special simulation.

I have two ideas to solve this problem:

- create a class for every combination of these parameter
- load the functions dynamic

For dynamic load csv is insufficient because the functions got parameter too e.g.`simulation.LTFArray.normal_weights(n, k)`.
We could use json instead which makes the experiment definition more verbose.

**Comment by MrM0nkey** [comment]
The two problems where fixed by a new version of the [architecture](https://cacoo.com/diagrams/ej0IjbAIwRJOV9Hd#C38C9).

**Comment by MrM0nkey** [comment]
@Chrism7 please help to fix the trainingset problem.

**Comment by MrM0nkey** [comment]
The fix for the trainingset problem leads to pass the puf simulation instance to the learner in order to generate crps in individual favor.

**Comment by nils-wisiol** [comment]
fixed by #19 and others
